js/tji-datasets-explore.js

```diff
@@ -1,15 +1,349 @@
 
+//TODO: make color constant object
+//WRITE documentation for how you expect people to use these classes
 
+var COLOR_TJI_BLUE = '#0B5D93';
+var COLOR_TJI_RED = '#CE2727';
+var COLOR_TJI_DEEPBLUE = '#252939';
+var COLOR_TJI_PURPLE = '#4D3B6B';
+var COLOR_TJI_YELLOW = '#F1AB32';
+var COLOR_TJI_TEAL = '#50E3C2';
+var COLOR_TJI_DEEPRED = '#872729';
+var COLOR_TJI_DEEPPURPLE = '#2D1752';
 
-// MAIN AJAX CALL TO ENDPOINT
 
+// *******************************************************************
+// * "Class" for a single-variable bar chart
+// *
+// * Shows counts of records grouped by a particular column.
+// *
+// * Constructor arguments:
+// *   elt_id: id of HTML canvas element to build Chart on
+// *   groupBy: column to group by
+// *   data: list of objects representing records
+// *   missing_data_label: stand-in label for records missing
+// *                       the groupBy column.
+// *******************************************************************
 
-//TOGGLE CHECKBOXES
-jQuery(document).ready(function($) {
-  $("#filter-section input:checkbox").prop("checked", true);
 
-  $('.js-filter-input-all').click(function() {
-      var $checkboxes = $(this).parent().parent().parent().find('input[type=checkbox]');
-      $checkboxes.prop('checked', $(this).is(':checked'));
-  }); 
-});
\ No newline at end of file
+var TJIGroupByBarChart = function(elt_id, groupBy, data, missing_data_label) {
+  this.elt_id = elt_id;
+  this.groupBy = groupBy;
+  this.chart = null;
+  this.missing_data_label = missing_data_label || '(not given)';
+  this.color_mapping = null;
+  this.create(data);
+}
+
+TJIGroupByBarChart.prototype.type = 'bar';
+
+TJIGroupByBarChart.prototype.color_palette = [COLOR_TJI_BLUE];
+
+TJIGroupByBarChart.prototype.create = function(data) {
+  var that = this;
+  var grouped = this.get_group_counts(data);
+
+  // 'create' is only run with the full data set.
+  // So we store our color mapping (e.g. "Male" = blue) so that
+  // subsequent filterings and changes don't alter the color
+  // associated with males.
+  this.color_mapping = {};
+  _.each(grouped.keys, function(k, idx) {
+    that.color_mapping[k] = that.color_palette[idx % that.color_palette.length];
+  });
+  var colors = _.map(grouped.keys, function(k) {
```

Comment:
Just a thought, but it might make sense to always show all the colors/grouped.keys in the legend for any given doughnut chart, so it's more obvious when a group is not represented within the data. For ex, if I deselected a particular year, and suddenly in the race doughnut all hispanic people disappeared, it might not be obvious if Hispanic is not shown in the legend.

Reply:
This would also slim down the update and create code, which is kinda nice.

Reply:
On it